### DISCUSSION TITLE
Uploading config.json via HTTP

### DISCUSSION THREAD

**Discussion by markusressel** [HASwitchPlate/openHASP#364] [Q&A] [answered]
Hey there, this is a really awesome project, thx!

To make it a bit easier for myself to manage the configuration of the plates in my home, I wrote a small helper to apply some preprocessing to the `*.jsonl` files and upload them to a plate in one single command.

To upload files I took a look at how the webserver saves changes made in the browser and replicated that HTTP request. This works pretty well so far, **except for the** `config.json`.

Uploading a file to the `/config.json` path succeeds, but the resulting file on the device was completely empty, resulting in what is essentially a factory reset after rebooting the plate.

### So the Question is:
Is there a way to upload a modified `config.json`? 
Or is it not possible on purpose?
If so it should probably return an error instead of writing an empty file :sweat_smile: 

I saw that there is ongoing progress on the HTTP website itself ( #246 ) and also for implementing an FTP server ( #283 ), I guess both of these should be linked to this question.

### Links
You can find the source to my tool here: [openhasp-config-manager](https://github.com/markusressel/openhasp-config-manager)

**Comment by fvanroie** [comment, marked as answer]
The `config.json` is read into memory at boot and overwritten from the _running_ configuration at reboot.
There is no code to handle uploads of `config.json` like that. It is only used to temporarily store the settings between reboots. It wasn't intended to be changed *on-the-fly*. Use the api calls in the web server to emulate the Configuration forms.

In the future this file will likely disappear from the filesystem. The settings should be moved to the NVS partition to prevent reading/writing it like that. The FTP server could now read sensitive contents of `config.json`, which shouldn't be possible...

**Reply by markusressel** [reply]
> Use the api calls in the web server to emulate the Configuration forms.

Alright, should be doable :+1:
Thx for the info :heart:

**Reply by fvanroie** [reply]
There is currently no `/api/config/` endpoint to get/set the *whole* config at once, but that's a future possibility.
Also, you can use the `config/*` commands over mqtt to get/set the configuration too.
